TraceQL: Experimental "not structural" operators (#2993)
grafana/tempo: docs/sources/tempo/traceql/_index.md

### Structural

These spanset operators look at the structure of a trace and the relationship between the spans. Structural operators ALWAYS return
matches from the right hand side of the operator.

- `{condA} >> {condB}` - The descendant operator (`>>`) looks for spans matching `{condB}` that are descendants of a span matching `{condA}`
- `{condA} << {condB}` - The ancestor operator (`<<`) looks for spans matching `{condB}` that are ancestor of a span matching `{condA}`
- `{condA} > {condB}` - The child operator (`>`) looks for spans matching `{condB}` that are direct child spans of a parent matching `{condA}`
- `{condA} < {condB}` - The parent operator (`<`) looks for spans matching `{condB}` that are direct parent spans of a child matching `{condA}`

> - `{condA} ~ {condB}` - The sibling operator (`~`) looks that spans matching `{condB}` that have at least one sibling matching `{condA}`.

Reviewer: Should this be looks at spans instead of looks that? 
```suggestion
- `{condA} ~ {condB}` - The sibling operator (`~`) looks at spans matching `{condB}` that have at least one sibling matching `{condA}`.
```

For example, to find a trace where a specific HTTP API interacted with a specific database:

```
{ span.http.url = "/path/of/api" } >> { span.db.name = "db-shard-001" }
```

> ### Experimental Structural

Reviewer: I added the standard wording that we use for an experimental feature or capability. What do you think? 
```suggestion
### Experimental structural

{{% admonition type="warning" %}}
These experimental spanset operators is an [experimental feature](/docs/release-life-cycle/). Engineering and on-call support is not available. Documentation is either limited or not provided outside of code comments. No SLA is provided.  
{{% /admonition %}}
```

Author: I don't mind adding a warning, but these are OSS docs. I don't think it makes sense to talk about SLAs or support.

> These spanset operators look at the structure of a trace and the relationship between the spans. They are marked experimental b/c they
> are known to sometimes return false positives. However, they can be very useful and we encourage users to try them and give feedback.

Reviewer: We say that these operators can be useful. Do we need to elaborate more on that or are the provided examples enough? 
```suggestion
These spanset operators look at the structure of a trace and the relationship between the spans. They are marked experimental because they
are known to sometimes return false positives. However, they can be very useful (see examples below). We encourage users to try them and give feedback.
```

> - `{condA} !>> {condB}` - The not descendant operator (`!>>`) looks for spans matching `{condB}` that are not descendant spans of a parent matching `{condA}`
> - `{condA} !<< {condB}` - The not ancestor operator (`!<<`) looks for spans matching `{condB}` that are not ancestor spans of a child matching `{condA}`
> - `{condA} !> {condB}` - The not child operator (`!>`) looks for spans matching `{condB}` that are not direct child spans of a parent matching `{condA}`
> - `{condA} !< {condB}` - The not parent operator (`!<`) looks for spans matching `{condB}` that are not direct parent spans of a child matching `{condA}`
> - `{condA} !~ {condB}` - The not sibling operator (`!~`) looks that spans matching `{condB}` that do not have at least one sibling matching `{condA}`.

Reviewer: Suggesting hyphens for the compound adjectives. The words like "not-descendant" and "not-ancestor" function as descriptors for operator, so I've made them compound adjectives by adding the hyphen. 
```suggestion
- `{condA} !>> {condB}` - The not-descendant operator (`!>>`) looks for spans matching `{condB}` that are not descendant spans of a parent matching `{condA}`
- `{condA} !<< {condB}` - The not-ancestor operator (`!<<`) looks for spans matching `{condB}` that are not ancestor spans of a child matching `{condA}`
- `{condA} !> {condB}` - The not-child operator (`!>`) looks for spans matching `{condB}` that are not direct child spans of a parent matching `{condA}`
- `{condA} !< {condB}` - The not-parent operator (`!<`) looks for spans matching `{condB}` that are not direct parent spans of a child matching `{condA}`
- `{condA} !~ {condB}` - The not-sibling operator (`!~`) looks that spans matching `{condB}` that do not have at least one sibling matching `{condA}`.
```

For example, to find a trace with a leaf span in the service "foo":

```
{ } !< { resource.service.name = "foo" }
```

To find a span that is the last error in a series of cascading errors:

```

> { status != error } !< { status = error }

Reviewer 2: Is this example 100% right?  I think the both clauses should be `status = error`, and then the query would be "find errors that aren't parents of other errors"

Author: good catch. fixing.
